parlai/tasks/natural_questions/agents.py

```diff
@@ -0,0 +1,275 @@
+    """
+    context_text = example[_context_type_key(is_html=True)].encode()
+    candidate_long_answers = []
+    for long_answer_span in example['long_answer_candidates']:
+        start_index = long_answer_span['start_byte']
+        end_index = long_answer_span['end_byte']
+        answer = context_text[start_index:end_index].decode()
+        candidate_long_answers.append(answer)
+    return candidate_long_answers
+
+
+def _create_long_answer_from_span_text(simplified_example):
+    """
+    Creates a list of long answer candidates, from their spans on the document.
+
+    This function gets the full article from the input simplified example
+    dictionary (using key 'document_text'), then iterates through the long
+    answer spans (from 'long_answer_candidates' key) and creates a list of
+    slices from the article, using the 'start_token' and 'end_token' values in
+    the list of long answer candidate spans.
+
+    Returns a list of long answers. Each long answer is a substring from the
+        simplified HTML text.
+
+    :param simplified_example: a dict that contains one simplified example/entry
+        from NQ dataset.
+    """
+    context_text = simplified_example[_context_type_key(is_html=False)]
+    candidate_long_answers = []
+    splitted_tokens = context_text.split(' ')
+    for long_answer_span in simplified_example['long_answer_candidates']:
+        start_index = long_answer_span['start_token']
+        end_index = long_answer_span['end_token']
+        answer = ' '.join(splitted_tokens[start_index:end_index])
+        candidate_long_answers.append(answer)
+    return candidate_long_answers
+
+
+class NaturalQuestionsTeacher(ChunkTeacher):
+    """
+    The base teacher class for Natural Questions dataset challenge.
+
+    This class implements the core functionalities for other teachers. The other four
+    variations of teachers are made by setting two object attributes (use_html,
+    use_long_answer) to either True or False.
+    """
+
+    def __init__(self, opt, shared=None):
+        build(opt)
+        self.use_html = opt.get('use_html', False)
+        self.use_long_answer = opt.get('use_long_answer', False)
+        self.id = 'natural_questions'
+        self.opt = copy.deepcopy(opt)
+        self.dtype = self.opt['datatype'].split(':')[0]
+        self.dpath = os.path.join(self.opt['datapath'], DATASET_NAME_LOCAL, self.dtype)
+        super().__init__(self.opt, shared)
+
+    def _simplify(self, example):
+        if self.use_html:
+            return example
+        return simplify_nq_example(example)
+
+    def _get_data_folder(self):
+        return self.dpath
+
+    def get_fold_chunks(self, opt) -> List[int]:
+        if 'train' == self.dtype:
+            return list(range(50))
+        elif 'valid' == self.dtype:
+            return list(range(5))
+        raise ValueError(f'Invalid data type: "{self.dtype}"')
+
+    def get_num_samples(self, opt) -> Tuple[int, int]:
```

Comment:
Should I simply hardcode this with an integer?

Reply:
no but it might be better to cache the result in an attr `self.n_samples`

Reply:
note this is still unresolved

Reply:
Sorry, I missed this.